// voerka-discover/config.default.js
module.exports = {
    cloud: {
        enabled: true,
        debug: true,
        broker: 'tcp://192.168.116.196:1885',
        username: '',
        password: '',
        willDebug: true,
        inviteTopic: '/voerka/discover/invite',
        confirmTopic: '/voerka/discover/+/answer',
        configTopic: '/voerka/discover/{SN}/config',
        serialize: 'lib/voerkaTransfer/protocols/serializers/msgPackSerialize',
        deserialize: 'lib/voerkaTransfer/protocols/serializers/msgPackDeserialize'
    },
    local: {
        enabled: true,
        debug: false,
        multicast: {
            port: 3888,
            addr: '239.8.8.8'
        },
        udpPort: 3888,
        deviceUdpPort: 2888,
        interval: '2s',
        serialize: 'lib/voerkaTransfer/protocols/serializers/msgStructBufferSerialize',
        deserialize: 'lib/voerkaTransfer/protocols/vimp/serializers/msgStructBufferDeserialize'
    },
    //Protocolos de descubrimiento de terceros extensibles, como qst5led
    timeout: '5min',
    pubInterval: '100ms'
};
